modules/table/src/main/java/org/apache/ignite/internal/utils/RebalanceUtil.java

```diff
@@ -0,0 +1,135 @@
+     * @param revision Revision of Meta Storage that is specific for the assignment update.
+     * @param metaStorageMgr Meta Storage manager.
+     */
+    public static void updateAssignmentsKeys(String partId, Collection<ClusterNode> baselineNodes, int partitions, int replicas,
+            long revision, MetaStorageManager metaStorageMgr) {
+        ByteArray partChangeTriggerKey = partChangeTriggerKey(partId);
+
+        ByteArray partAssignmentsPendingKey = partAssignmentsPendingKey(partId);
+
+        ByteArray partAssignmentsPlannedKey = partAssignmentsPlannedKey(partId);
+
+        ByteArray partAssignmentsStableKey = partAssignmentsStableKey(partId);
+
+        byte[] partAssignmentsBytes = ByteUtils.toBytes(AffinityUtils.calculateAssignments(baselineNodes, partitions, replicas));
+
+        //    if empty(partition.change.trigger.revision) || partition.change.trigger.revision < event.revision:
+        //        if empty(partition.assignments.pending) && partition.assignments.stable != calcPartAssighments():
+        //            partition.assignments.pending = calcPartAssignments()
+        //            partition.change.trigger.revision = event.revision
+        //        else:
+        //            if partition.assignments.pending != calcPartAssignments
+        //                partition.assignments.planned = calcPartAssignments()
+        //                partition.change.trigger.revision = event.revision
+        //            else
+        //                remove(partition.assignments.planned)
+        //    else:
+        //        skip
+        var iif = If.iif(or(exists(partChangeTriggerKey), value(partChangeTriggerKey).lt(ByteUtils.longToBytes(revision))),
+                If.iif(and(exists(partAssignmentsPendingKey), value(partAssignmentsStableKey).ne(partAssignmentsBytes)),
+                        ops(
+                                put(partAssignmentsPendingKey, partAssignmentsBytes),
+                                put(partChangeTriggerKey, ByteUtils.longToBytes(revision))
+                        ).yield(),
+                        If.iif(value(partAssignmentsPendingKey).ne(partAssignmentsBytes),
+                                ops(
+                                        put(partAssignmentsPlannedKey, partAssignmentsBytes),
+                                        put(partChangeTriggerKey, ByteUtils.longToBytes(revision))
+                                ).yield(),
+                                ops(remove(partAssignmentsPlannedKey)).yield())),
+                ops().yield());
+
+        metaStorageMgr.invoke(iif);
+    }
+
+    /**
+     * Key that is needed for the rebalance algorithm.
+     *
+     * @param partId Unique identifier of a partition.
+     * @return Key for a partition.
+     * @see <a href="https://github.com/apache/ignite-3/blob/main/modules/table/tech-notes/rebalance.md">Rebalnce documentation</a>
+     */
+    public static ByteArray partChangeTriggerKey(String partId) {
```

Comment:
`tableId` must be a part of key too, wdyt? Here and in the other keys.

Reply:
In java doc we say that this partId must be unique id of a partition. If you check the usage of this method, you can see that we call `partitionRaftGroupName(UUID tblId, int partition)`, that uses tableId

Reply:
Yep, I see